G Suite Security Alert Center Integration

crestdatasystems opened the pull request:
<!-- REMINDER: THIS IS A PUBLIC REPO DO NOT POST HERE SECRETS/SENSITIVE DATA -->

## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related issues:
fixes: https://github.com/demisto/etc/issues/19046


## Description
G Suite Security Alert Center allows users to fetch different alert types such as Suspicious login, Device compromised, Leaked password, and more. Users can delete or undelete a single alert or batch of alerts and retrieve the alert's metadata. This integration allows users to provide feedback for alerts and fetch existing feedback for a particular alert.
    
## Minimum version of Demisto
- [x] 5.0.0
- [ ] 5.5.0
- [ ] 6.0.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/demisto/content?pullRequest=11435) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/demisto/content?pullRequest=11435) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: parimalje<br/>:white_check_mark: crestdatasystems<br/>:white_check_mark: crestsantoshgiri<br/>:x: meetvalani_crest<br/><hr/>**meetvalani_crest** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/demisto/content?pullRequest=11435) it.</sub>

reutshal commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.yml:
```diff
@@ -0,0 +1,805 @@
+  name: isFetch
+  required: false
+  type: 8
+- defaultvalue: G Suite Security Alert Center
```
Can we supply pre-defined  types list here?

crestdatasystems replied:
Will check if multi-select with adding custom value is possible or not

crestdatasystems replied:
Added the multi-select option in alert_type parameter with support for adding custom values

reutshal commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.yml:
```diff
@@ -0,0 +1,805 @@
+      type: String
+    - contextPath: GSuiteSecurityAlert.Alert.data.messages.attachmentsSha256Hash
+      description: The SHA256 hash of email's attachment and all MIME parts.
+      type: Unknown
```
Please change the type

crestdatasystems replied:
Array is returned as part of  `attachmentsSha256Hash` field that's why we have kept type as Unknown

reutshal replied:
OK, so please mention in the description that this is an array

crestdatasystems replied:
Will change as required.

reutshal commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.yml:
```diff
@@ -0,0 +1,805 @@
+  script: '-'
+  subtype: python3
+  type: python
+tests:
```
Please attach the test playbook here

crestdatasystems replied:
Will change as required.

crestdatasystems replied:
Added as required

reutshal reviewed (changes requested):
Thanks! 
When can we do a demo?

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+                alert_filter += f' OR type="{alert_types[index]}"'
+
+    # If next_page_token is present replace filter with filter present in last_run as API does not support
+    # change in filter.
```
We have kept this logic because whenever we have `next_page_token` from an API call and we pass that param to fetch next page data with changed filter param value, the server responds with "Invalid Argument" `400` error. 

So to handle this in fetch-incident we have kept the logic that until next_page_token is present use the previous filter, if next_page_token is not present the latest filter value would be used from configuration.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+        alert_filter += f' AND {advance_filter}'
+
+    if 'type' not in alert_filter:
+        alert_types = params.get('alert_type', '').split(',')
```
The additional Info of `filter` param contains following text: `Note: Takes higher precedence against filter parameters.`

We can still mention in `alert_type` param that it will be overridden if type is present in `filter` param. 

@reutshal Kindly confirm if we should mention it or not.

reutshal replied:
Yes I think that will be best

crestdatasystems replied:
Will change as required.

crestdatasystems replied:
Did not require to change this as added multi_select option as part of this param

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+            if index == 0 and alert_types[index] != '':
+                alert_filter += f' AND type="{alert_types[index]}"'
+            elif alert_types[index] != '':
+                alert_filter += f' OR type="{alert_types[index]}"'
```
Sure!

crestdatasystems replied:
Will check if parenthesis is supported and check if filter works correctly.

crestdatasystems replied:
Added the parenthesis

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+    list_alert_feedback_response = gsuite_client.http_request(
+        url_suffix=URL_SUFFIX['FEEDBACK'].format(args['alert_id']),
+        method='GET',
+        params=GSuiteClient.remove_empty_entities(params))
```
Yes! It brings all the feedbacks

reutshal commented:
@kirbles19 @ShirleyDenkberg This is a pack supported by us, need your help with the documentation review 🙏 
If you need me to point to the specific files please DM me and we will go over it.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/pack_metadata.json:
```diff
@@ -0,0 +1,17 @@
+{
+    "name": "G Suite Security Alert Center",
+    "description": "G Suite Security Alert Center allows users to fetch different alert types such as Suspicious login, Device compromised, Leaked password, and more. Users can delete or undelete a single alert or batch of alerts and retrieve the alert's metadata. This integration allows users to provide feedback for alerts and fetch existing feedback for a particular alert.",
+    "support": "partner",
```
@reutshal Do we have to add support value as `xsoar` or `XSOAR`?

reutshal replied:
The one I suggested :)

crestdatasystems replied:
Will change as required.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+    return success_list, failed_list
+
+
+def check_required_arguments(required_arguments: List[str], args: Dict[str, Any]):
```
We have implemented this method because when we pass only whitespace in required argument the Cortex XSOAR allows the value and does not show required argument not present. Once we fetch the command arguments in code we use `GSuiteClient.strip_dict()` method which strips the value of all arguments and removes empty arguments. 

So, to check if all the required arguments are present before hitting the API, arguments are passed through this method.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+
+    # Fetch createTime of latest alert
+    if len(alerts) > 0:
+        timestamp = alerts[-1]['createTime']
```
While validating params for fetch incidents in `validate_params_for_fetch_incidents` method we have added `orderBy` key so that while fetching alerts from Alert centre the alerts come in sorted manner. 
![image](https://user-images.githubusercontent.com/60967033/108345675-90d13f00-7204-11eb-84b2-3cf4452651c8.png)

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/pack_metadata.json:
```diff
@@ -0,0 +1,17 @@
+    "description": "G Suite Security Alert Center allows users to fetch different alert types such as Suspicious login, Device compromised, Leaked password, and more. Users can delete or undelete a single alert or batch of alerts and retrieve the alert's metadata. This integration allows users to provide feedback for alerts and fetch existing feedback for a particular alert.",
+    "support": "partner",
+    "currentVersion": "1.0.0",
+    "author": "XSOAR",
```
Will change as required.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/pack_metadata.json:
```diff
@@ -0,0 +1,17 @@
+    "description": "G Suite Security Alert Center allows users to fetch different alert types such as Suspicious login, Device compromised, Leaked password, and more. Users can delete or undelete a single alert or batch of alerts and retrieve the alert's metadata. This integration allows users to provide feedback for alerts and fetch existing feedback for a particular alert.",
+    "support": "partner",
+    "currentVersion": "1.0.0",
+    "author": "XSOAR",
```
Will change as required.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.yml:
```diff
@@ -0,0 +1,805 @@
+  name: alert_type
+  required: false
+  type: 0
+- additionalinfo: |-
```
Will change as required.

crestdatasystems replied:
Added the note

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+        alert_filter += f' AND {advance_filter}'
+
+    if 'type' not in alert_filter:
+        alert_types = params.get('alert_type', '').split(',')
```
Will change as required.

crestdatasystems replied:
Did not require to change this as added multi_select option as part of this parameter

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+    Prepare the Human readable info for alerts command.
+
+    :param alerts: The alerts data.
+    :param header: Header of the hr.
```
Will change as required.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+@logger
+def gsac_get_alert_command(client, args: Dict[str, str]) -> CommandResults:
+    """
+    Get alert from G Suite Security Alert center.
```
Will change as required.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+        method='GET',
+        params=GSuiteClient.remove_empty_entities(params))
+
+    no_records = len(list_alert_feedback_response.get('feedback', []))
+    if no_records == 0:
```
Will change as required.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+            if len(feedback_response) > 0:
+                # Fetch latest feedback
+                alert['feedback'] = feedback_response[0]
+        incident = {
```
Will change as required.

crestdatasystems commented on Packs/GSuiteSecurityAlertCenter/Integrations/GSuiteSecurityAlertCenter/GSuiteSecurityAlertCenter.py:
```diff
@@ -0,0 +1,719 @@
+        # This is the call made when pressing the integration Test button.
+        if demisto.command() == 'test-module':
+            result = test_module(gsuite_client, {}, params)
+            demisto.results(result)
```
Will change as required.

crestdatasystems commented:
@reutshal  Done the following changes as requested, kindly review!

- Updated the following values in pack_metadata.json:
    "support": "xsoar"
    "author": "Cortex XSOAR"
- Added a note for Filter that " To fetch alerts using createTime, use the first fetch time interval parameter."
- Added occurred key for alert create time in fetch-incident.
- Updated the note for Alert Type parameter. 
- Updated the note for Filter parameter.
- Updated the Alert Type parameter type to multi-select
- Updated the 'undelete' command name to 'recover'
- Added the page_size argument for alert-feedback-list whose default value is 50
- Updated the descriptions for gsac-alert-delete & gsac-alert-recover & gsac-alert-list commands.
-  Added all other code best practice related changes
- Added test_playbook in tests key of yml
-

ShirleyDenkberg commented:
@reutshal @EliorKedar Doc review completed.
